assets/in

```diff
@@ -8,6 +8,77 @@ exec 1>&2 # redirect all output to stderr for logging
 
 source $(dirname $0)/common.sh
 
+
+git_shallow_clone() {
+  local branch="$1" ref="$2" depth="$3" repo_url="$4" dest_dir="$5"
+
+  declare -a git_clone_opts
+  if [ -n "$branch" ]; then
+    git_clone_opts+=(--branch "$branch")
+  fi
+
+  if test "${depth:-0}" -gt 0; then
+    git_clone_opts+=(--depth "$depth")
+  fi
+
+  git clone --single-branch "${git_clone_opts[@]}" "$repo_url" "$dest_dir"
+
+  # A shallow clone may not contain the Git commit $ref:
+  # 1. The depth of the shallow clone is measured backwards from the latest
+  #    commit on the given head (master or branch), and in the meantime there may
+  #    have been more than $depth commits pushed on top of our $ref.
+  # 2. If there's a path filter (`paths`/`ignore_paths`), then there may be more
+  #    than $depth such commits pushed to the head (master or branch) on top of
+  #    $ref that are not affecting the filtered paths.
+  #
+  # In either case we try to deepen the shallow clone until we find $ref, reach
+  # the max depth of the repo, or give up after a given depth and resort to deep
+  # clone.
+  if [ "$depth" -gt 0 ]; then
+    local max_depth=128
+
+    d="$depth"
+    while ! git -C "$dest_dir" checkout -q "$ref" &>/dev/null; do
+      # once the depth of a shallow clone reaches the max depth of the origin
+      # repo, Git silenty turns it into a deep clone
+      if [ ! -e "$dest_dir"/.git/shallow ]; then
+        echo "Reached max depth of the origin repo while deepening the shallow clone, it's a deep clone now"
+        break
+      fi
+
+      echo "Could not find ref ${ref} in a shallow clone of depth ${d}"
+
+      (( d *= 2 ))
+
+      if [ "$d" -gt "$max_depth" ]; then
+        echo "Reached depth threshold ${max_depth}, falling back to deep clone..."
+        git -C "$dest_dir" fetch --unshallow origin
+
+        break
+      fi
+
+      echo "Deepening the shallow clone to depth ${d}..."
+      git -C "$dest_dir" fetch --depth "$d" origin
+    done
+  fi
+
+  git -C "$dest_dir" checkout -q "$ref"
+}
+
+clone_submodule() {
+  local submodule="$1" depth="$2" dest_dir="$3"
+
+  echo "Cloning submodule '$submodule'..."
+
+  local submodule_branch submodule_ref submodule_uri
+  submodule_branch="$(git config --get "submodule.${submodule}.branch" || true)"
+  submodule_ref="$(git ls-tree -d HEAD "$submodule" | awk '{ print $3 }')"
+  submodule_uri="$(git config --get "submodule.${submodule}.url")"
+
+  git_shallow_clone "${submodule_branch:-master}" "$submodule_ref" "$depth" "$submodule_uri" "$dest_dir/$submodule"
```

Comment:
This is interesting. Is this effectively doing a `git clone` to the submodule path? Does that leave the tree clean, equivalent to `git submodule update --init`?